compiler/rustc_errors/src/diagnostic_builder.rs

```diff
@@ -305,9 +305,7 @@ impl<'a, G: EmissionGuarantee> DiagnosticBuilder<'a, G> {
             }
         };
 
-        if dcx.inner.lock().flags.dont_buffer_diagnostics
-            || dcx.inner.lock().flags.treat_err_as_bug.is_some()
-        {
+        if dcx.inner.lock().flags.treat_err_as_bug.is_some() {
```

Comment:
Does it even serve a purpose to fire `treat_err_as_bug` for diagnostics that are created, for example, then canceled? Shouldn't we be checking `treat_err_as_bug` on the emission path?

Reply:
It's a fair question. There's a lot of special diagnostics cases (e.g. for other `-Z` options) that are dubious, because there are lots of different possible chokepoints for special diagnostic behaviour: emit, cancel, stash, buffer, into_diagnostic, etc. (And don't ask about how errors and warnings are counted, yikes, it's a mess.) I think people have just inserted them one by one over the years without necessarily having a full picture of the diagnostics system. I'm working my way through these, improving them as much as I can.

I can certainly give consideration to your question when I consider what to do with `treat_err_as_bug`, but I think that can wait until after this PR is merged. Indeed, this PR is exactly the kind of small cleanup that then smooths the way for other cleanups, like making `treat_err_as_bug` nicer.